### ISSUE TITLE
Incompatible with Bolt 3.2.2 & 3.2.3

### ISSUE THREAD

**Issue by graham73may** [CondeNastDigital/cnBolt-Extensions-RelationList#9]
I've just done a Bolt composer update to 3.2.3 and have been experiencing an issue that meant I was unable to save if the edit page had a repeater field. (The issue still occurred even if a relationList field was not present - see below). 

Pages for example, without the plugin works fine, but with relationList installed I get the following errors. 

![image](https://cloud.githubusercontent.com/assets/6642498/20725552/1fbf3510-b66a-11e6-96e2-704676e25ed5.png)

The issue is being thrown from `/vendor/bolt/bolt/app/src/js/widgets/fields/fieldRepeater.js::150` so I'm guessing there's some sort of compatibility issue that's been introduced in the latest release?

My pages is defined as
```
__nodes:
    # Page Parent
    parent: &parent
        parent:
            group: meta
            label: "Page parent"
            type: select
            values: page/title

# Pages
pages:
    class: Soapbox\Entities\Content
    name: Pages
    singular_name: Page
    fields:
        title:
            class: large
            group: content
            required: true
            type: text
        slug:
            type: slug
            uses: title
            route_prefix: "/"
    # Defaults
        excerpt:
            height: 100px
            type: textarea
        body:
            group: content
            height: 300px
            type: html
    # Page Links
        page_links:
            group: content
            label: "Links block"
            type: repeater
            fields:
                link_text:
                    label: "Link text"
                    type: text
                link:
                    label: "Link"
                    error: "Please enter a valid URL."
                    type: text
    # Hierarchy selection
        <<: *parent
    # Template selection
        template:
            group: meta
            filter: '*.twig'
            type: templateselect
    taxonomy: [ tags ]
    hierarchical: true
    icon_many: "fa:file"
    icon_one: "fa:file"
    recordsperpage: 100
```

**Comment by graham73may** [comment]
Courtesy of @mrenigma on Bolt Slack 


Okay so it looks like `var init = {}` in init.js might be getting overwritten. 

Yep being overwritten by relationlist:

```
that.init = function(options) {

        if(typeof(SirTrevor)) {
            Object.keys(options).forEach(function (block) {

                if (!(options[block] instanceof Object && options[block].hasOwnProperty('type') && options[block].type == 'relationlist'))
                    return;

                var newBlock = {
                    type: block,
                    custom: options[block]
                };

                newBlock.custom.subFieldName = block;
                newBlock.custom.contenttype = $('[name="contenttype"]').val();

                if (typeof(SirTrevor.Blocks[block]) === 'undefined') {
                    newBlock = jQuery.extend({}, that.blockPrototype, newBlock);
                    SirTrevor.Blocks[block] = SirTrevor.Block.extend(newBlock);
                }
            });
        }
    };
```

In sir-trevor.js lines 87-109

**Comment by graham73may** [comment]
Suggested quick-fix from @Ross on Slack

> I think it's an optional plugin for relationlist
> as far as I can see, if you just blank the file in public/extensions/cnd/relationlist/js/extension-for/sir-trevor.js nothing bad happens.
> and things will work as expected until a fix is available

**Comment by ross** [comment]
@graham73may I believe you have the wrong Ross again, /cc @rossriley

**Comment by graham73may** [comment]
Can't believe I did it again, apologies. I didn't think it would connect to you as it didn't give me the username select pop-up. I'll be more careful in the future!

**Comment by ToBe998** [comment]
You are right, the *sir-trevor* part of this extension is an optional part which should not affect you in any way. We also have a (not yet) public extension that provides a new field based on the Sir Trevor editor. The optional part provides a RelationList Block for this Sir Trevor editor.
Since you dont have this Sir Trevor extension, that part should not have any impact on you.

Problem is, we are not using the repeater fields here since they do not offer the flexibility we need. A small side effect might be that the RelationList extension is not tested againt repeater fields. I will forward your proposed fixes to the main dev on the extension and we can hopefully integrate it.

You can also allways provide a PR for a change and i will happily merge it if it doesnt break anything.

(And a ping to @rossralev, the current main dev on this extension)

**Comment by graham73may** [comment]
Hi @tobe998, thanks for the response. 

I'll provide a PR if I get a chance and can get my head around the code, I'm a bit worried about breaking the extension in this instance if I'm honest, especially as it connects with one of yours that's not public. 

If you are able to provide a fix in the future that would be great, it's an awesome field type!

**Comment by ToBe998** [comment]
Thanks! THis extensions is actually the result of "not reading the bolt manual" :)
We didnt know you could use select fields using other contenttypes back then. But it has grown into something we are using on several sites now. :)

**Comment by ToBe998** [comment]
@rossralev is also on bolt slack and might have time to discuss the sir-trevor part. It's his contribution to the etension, so he might know best what to do. (Im more of a backend guy :P )

**Comment by graham73may** [comment]
The reason I'm using it is as a field that allows the user to select content from more than one contenttype. Like a multi-contenttype select field, from what I'm aware it's not possible to do that with stock Bolt. (If you know differently please do tell!) 

This field is really vital for related content / featured content entry!

**Comment by ToBe998** [comment]
You are right, thats the main reason we have this extension. There also is a selectContent twig function that can fetch more than one contenttype at once and filter sort by various criteria, including taxonomy. Not public yet, but a hot candidate for it.

**Comment by rossralev** [comment]
Hi, 

Tomorrow i will take a look, why this problem appears. And publish a fix. 
Can you also gives an overview of what other extensions you have installed

Regards 
Rosen

**Comment by graham73may** [comment]
Hi @rossralev, thanks, that's great to hear! 

The only other extensions I have are menu editor and conditional fields. 

I have a number of local extensions but none of these add any JavaScript assets.

**Comment by rossralev** [comment]
Hi @graham73may 

I've just upgraded my local version to 3.2.3 and removed the sir-trevor extension that i had. It still works. Can you send me your extensions/content.yml 
I have to reproduce your system to be able to understand what is exactly going on. 

The that.init is part of an Object. It should not affect any other core functionalities (one never knows though :) )

**Comment by graham73may** [comment]
Hi @rossralev 

I was able to replicate the issue by:

1. installing a brand new bolt 3.2.3, 
1. adding the relationlist extension 
1. navigating to the edit page of a contenttype with a repeater. 

Here's my contenttypes.yml from a test site I was trying to debug this on with the issue occurring:

```
__nodes:
    # Page Parent
    parent: &parent
        parent:
            group: meta
            label: "Page parent"
            type: select
            values: page/title


# Pages
pages:
    class: Soapbox\Entities\Content
    name: Pages
    singular_name: Page
    fields:
        title:
            class: large
            group: content
            required: true
            type: text
        slug:
            type: slug
            uses: title
            route_prefix: "/"
    # Defaults
        excerpt:
            height: 100px
            type: textarea
        body:
            group: content
            height: 300px
            type: html
    # Page Links
        page_links:
            group: content
            label: "Links block"
            type: repeater
            fields:
                link_text:
                    label: "Link text"
                    type: text
                link:
                    label: "Link"
                    error: "Please enter a valid URL."
                    type: text
    # Hierarchy selection
        <<: *parent
    # Template selection
        template:
            group: meta
            filter: '*.twig'
            type: templateselect
    taxonomy: [ tags ]
    hierarchical: true
    icon_many: "fa:file"
    icon_one: "fa:file"
    recordsperpage: 100

```

Thanks,

**Comment by rossralev** [comment]
Thank you, I will check it out and write back :) @graham73may

**Comment by rossralev** [comment]
Hi @graham73may 
can please check to set the plugin version to dev-master and update. 
I made the small fix. :)

**Comment by graham73may** [comment]
Hi @rossralev 

I've installed dev-master
![image](https://cloud.githubusercontent.com/assets/6642498/20749128/f0872e04-b6e8-11e6-9106-9a4dc4046f17.png)


But I'm still getting the same errors on repeaters:
![image](https://cloud.githubusercontent.com/assets/6642498/20749112/dcbc34d2-b6e8-11e6-8600-d1cd36acb49f.png)

**Comment by graham73may** [comment]
@rossralev 

The issue is caused by Bolt's `init` being overwritten by one inside `sir-trevor.js` here: 
https://github.com/CondeNastDigital/cnBolt-Extensions-RelationList/blob/master/web/js/extension-for/sir-trevor.js#L87

If I comment out the `that.init` block, everything works as normal. 

Here's where the exception is being thrown and what `init` equals
![image](https://cloud.githubusercontent.com/assets/6642498/20750710/214110bc-b6f0-11e6-81fd-f4554e82c364.png)

**Comment by rossralev** [comment]
Hi Again, @graham73may 

In sir-trevor.js, on line 115, you should be able to see:
 var relationListST = new RelationlistST({

I have overseen the lack of "new", so this is what i have added today. Can you please check if your public/extensions/vendor/cnd/relationlist/js/extension-for/sir-trevor.js has the new version.

My latest commit has an id:  5a381de. I see that your version is still at 3c107c.

**Comment by graham73may** [comment]
HI @rossralev 

Sorry you are correct, the `new` addition fixes the issue. 

The extensions store must not have updated yet with your latest commit. I installed `dev-master` through the Extend page -> show all versions.

**Comment by rossralev** [comment]
Super :) @graham73may 
Then I will prepare a new revision. Sorry for the inconvenience.

**Comment by ToBe998** [comment]
Fixed in 2.0.8. I also triggered an update on bolt's marketplace but it usually takes a couple of hours.
